posts/auditing-using-javers.md

```diff
@@ -0,0 +1,238 @@
+        ...
+    }
+}
+```
+
+We also excluded from the audit all the no-PHI data using the @DiffIgnore annotation
+
+Example :
+
+```java
+class MongoStoredEntity {
+
+    // ...
+
+    @DiffIgnore
+    @Property("description")
+    private String description
+
+    // ...
+}
+```
+
+# Retrieve the change
+As mentioned earlier, we stored all our auditing in a separate Mongo collection from client-facing data. However, we needed to find a way to display back the information collected. JaVers provides its' own JaVers Query Language (JQL), which can be used to query the Javers Repository for changes on a given class, object or property.
+
+Data can be retrieved from JaVers in 3 possible ways: Shadows, Changes and Snapshots.
+  * *Shadows* are historical versions of objects
+  * *Change* shows the difference between two objects
+  * *Snapshots* show the state of an object at a point in times
+
+JaVers uses the primary key on the table or collection in order to compile this information.
+
+##Shadows
+Shadows most easily offer a view of data history. Shadows are instances of your object from different points in time. Since Shadows are instances of your domain classes, they can be easily used.
+
+Example:
+```java
+Employee bob = new Employee("bob",
+                             1000,
+                             new Address("London"));
+      javers.commit("author", bob) ;      // initial commit
+
+      bob.setSalary(1200)  ;                // changes
+      bob.getPrimaryAddress().setCity("Paris");  //
+      javers.commit("author", bob);       // second commit
+```
+With the transactions above, we can query the following:
+```java
+Shadows shadows = javers.findShadows(QueryBuilder.byInstance(bob).build());
+```
+Shadows, as provided here is a wrapper object containing all the distinct instances of the Employee object bob.
+
+##Snapshots
+Snapshots are the historical state of a domain object captured as the property-value map. Snapshots are raw data stored in the JaversRepository. When an object is committed, JaVers makes a Snapshot of its state and persists it. JaVers reuses Snapshots and creates a new one only when a given object is changed.
+
+JaVers fetches snapshots in reversed chronological order.
+
+Looking once again at Bob, when we query the following:
+```java
+List<CdoSnapshot> snapshots = javers.findSnapshots( QueryBuilder.byInstance(bob).build());
+```
+CdoSnapshot, as seen here is a wrapper object containing the historical state of an object at a point in time.
+
+##Changes
+The Changes view of JaVers provides the list of differences between versions of a domain object. Changes are calculated as the difference between Snapshots loaded from the JaVers Repository. As per our client requirements, we needed to display the changes for an object from creation. We decided to use the **findChanges** method to do so. There are 3 types of changes tracked by JaVers: NewObject, ObjectRemoved and PropertyChange.
```

Comment:
```suggestion
The Changes view of JaVers provides the list of differences between versions of a domain object. Changes are calculated as the difference between snapshots loaded from the JaVers Repository. As per our client requirements, we needed to display the changes for an object from creation. We decided to use the **findChanges** method to do so. There are 3 types of changes tracked by JaVers: `NewObject`, `ObjectRemoved` and `PropertyChange`.
```

Reply:
Resolved! I would mark it as resolved, but that option isn't showing up anymore.